models/record.go

```diff
@@ -0,0 +1,273 @@
+//
+type RecordConfig struct {
+	Type             string            `json:"type"`   // All caps rtype name.
+	Name             string            `json:"name"`   // The short name. See above.
+	NameFQDN         string            `json:"-"`      // Must end with ".$origin". See above.
+	Target           string            `json:"target"` // If a name, must end with "."
+	TTL              uint32            `json:"ttl,omitempty"`
+	Metadata         map[string]string `json:"meta,omitempty"`
+	MxPreference     uint16            `json:"mxpreference,omitempty"`
+	SrvPriority      uint16            `json:"srvpriority,omitempty"`
+	SrvWeight        uint16            `json:"srvweight,omitempty"`
+	SrvPort          uint16            `json:"srvport,omitempty"`
+	CaaTag           string            `json:"caatag,omitempty"`
+	CaaFlag          uint8             `json:"caaflag,omitempty"`
+	TlsaUsage        uint8             `json:"tlsausage,omitempty"`
+	TlsaSelector     uint8             `json:"tlsaselector,omitempty"`
+	TlsaMatchingType uint8             `json:"tlsamatchingtype,omitempty"`
+	TxtStrings       []string          `json:"txtstrings,omitempty"` // TxtStrings stores all strings (including the first). Target stores only the first one.
+	R53Alias         map[string]string `json:"r53_alias,omitempty"`
+
+	Original interface{} `json:"-"` // Store pointer to provider-specific record object. Used in diffing.
+}
+
+// Copy returns a deep copy of a RecordConfig.
+func (rc *RecordConfig) Copy() (*RecordConfig, error) {
+	newR := &RecordConfig{}
+	err := copyObj(rc, newR)
+	return newR, err
+}
+
+// SetLabel sets the .Name/.NameFQDN fields given a short name and origin.
+// origin must not have a trailing dot: The entire code base
+//   maintains dc.Name without the trailig dot. Finding a dot here means
+//   something is very wrong.
+// short must not have a training dot: That would mean you have
+//   a FQDN, and shouldn't be using SetLabel().  Maybe SetLabelFQDN()?
+func (rc *RecordConfig) SetLabel(short, origin string) {
+	if strings.HasSuffix(origin, ".") {
+		panic(errors.Errorf("origin (%s) is not supposed to end with a dot", origin))
+	}
+	if strings.HasSuffix(short, ".") {
+		panic(errors.Errorf("short (%s) is not supposed to end with a dot", origin))
+		// NB(tlim): we should never get this panic.
+	}
+
+	short = strings.ToLower(short)
+	origin = strings.ToLower(origin)
+	if short == "" || short == "@" {
+		rc.Name = "@"
+		rc.NameFQDN = origin
+	} else {
+		rc.Name = short
+		rc.NameFQDN = dnsutil.AddOrigin(short, origin)
+	}
+}
+
+// SetLabelFQDN sets the .Name/.NameFQDN fields given a FQDN and origin.
+// fqdn may have a trailing "." but it is not required.
+// origin may not have a trailing dot.
+func (rc *RecordConfig) SetLabelFQDN(fqdn, origin string) {
+
+	if strings.HasSuffix(origin, ".") {
+		panic(errors.Errorf("origin (%s) is not supposed to end with a dot", origin))
+	}
+	if strings.HasSuffix(fqdn, "..") {
```

Comment:
There is a point where the user is just trying to shoot themselves in the foot. We don't check double dots in the middle for example. 

I feel like we should either do **more** rigorous validation here (`(\w+\.)+` maybe), or test fewer things.

In the future, I think maybe the validation package being separate is a mistake and we should just have a single call `Validate` on dnsconfig that cascades down to domains and records.

Reply:
The panics are only for assertions failures. i.e. the code is being used wrong.  I could change them to errors later, but these are literally "this should never happen" panics.

As far as validation... yes, that needs to be refactored some day.